typing/printtyp.ml

```diff
@@ -215,43 +224,9 @@ module S = String.Set
 let enabled = ref true
 let enable b = enabled := b
 
-(** Name mapping *)
-type mapping =
-  | Need_unique_name of int Ident.Map.t
-  (** The same name has already been attributed to multiple types.
-      The [map] argument contains the specific binding time attributed to each
-      types.
-  *)
-  | Uniquely_associated_to of Ident.t * out_name
-    (** For now, the name [Ident.name id] has been attributed to [id],
-        [out_name] is used to expand this name if a conflict arises
-        at a later point
-    *)
-  | Associated_to_pervasives of out_name
-  (** [Associated_to_pervasives out_name] is used when the item
-      [Stdlib.$name] has been associated to the name [$name].
-      Upon a conflict, this name will be expanded to ["Stdlib." ^ name ] *)
-
-let hid_start = 0
-
-let add_hid_id id map =
-  let new_id = 1 + Ident.Map.fold (fun _ -> Int.max) map hid_start in
-  new_id, Ident.Map.add id new_id  map
-
-let find_hid id map =
-  try Ident.Map.find id map, map with
-  Not_found -> add_hid_id id map
-
-let pervasives name = "Stdlib." ^ name
-
-let map = Array.make Namespace.size M.empty
-let get namespace = map.(Namespace.id namespace)
-let set namespace x = map.(Namespace.id namespace) <- x
-
-(* Names used in recursive definitions are not considered when determining
-   if a name is already attributed in the current environment.
-   This is a complementary version of hidden_rec_items used by short-path. *)
-let protected = ref S.empty
+(* Names bound in recursive definitions should be considered as bound
+   in the environment when printing identifiers *)
+let bound_in_recursion = ref M.empty
```

Comment:
My high-level explanation is that when printing recursive definitions there are two printing environment that are natural to consider: the one just before the recursive definitions and the one with the recursive components bound in the scope. The short-paths functions work more naturally on the `just-before` environment because the identifiers are not yet valid shorthand for longer names, 
```ocaml
module Avoid__me = struct
  type t = A
end
type u = Avoid__me.t
```
whereas the `name conflict` functions need to take in account the environment extended with the current recursive identifiers which are already in the environment:
```ocaml
type t = A
type u = [ `A of t ]
module M = struct
  type t = [ u | `B of t ]  
end
```

Similarly, only new definition can introduce new identifiers in the environment that

Concerning, indeed there is a bug in the printer in the case of recursive module where it is possible to shadow the recursive module being defined by a non-recursive one:
```
module rec M: sig
  type t
  type r = [`A of M.t]
  module M: sig
     type t
   end
  type s = [ `B | r ]
end = struct
  type t
  type r = [`A of M.t]
  module M = struct
     type t
   end
  type s = [ `B | r ]
end
```
Here the printer expand the definition of `s` into
```ocaml
    type s = [ `A of M.t | `B ]
```
which is refers to the shadowed module `M` directly.

Overall, I think that the best fix is to switch to a second explicit environment rather than rely on the ad-hoc `bound_in_recursion` type.

Reply:
It's okay if the code is not completely correct on first change -- incremental improvements are fine. But it would be nice to have comments (as you just wrote) to explain what the code is trying to be doing.